community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/TicketedProcessing.java

```diff
@@ -0,0 +1,185 @@
+    {
+        this.processor = processor;
+        this.executor = new DynamicTaskExecutor<>( 1, maxProcessors, maxProcessors, park, name,
+                threadLocalStateSupplier );
+        this.healthCheck = executor::assertHealthy;
+        this.processed = new ArrayBlockingQueue<>( maxProcessors );
+    }
+
+    public void submit( long ticket, FROM job )
+    {
+        submittedTicket.incrementAndGet();
+        executor.submit( threadLocalState ->
+        {
+            TO result = processor.apply( job, threadLocalState );
+            await( myTurnToAddToProcessedQueue, ticket, healthCheck, park );
+            while ( !processed.offer( result ) );
+            processedTicket.incrementAndGet();
+        } );
+    }
+
+    /**
+     * Essentially starting a thread {@link #submit(long, Object) submitting} a stream of inputs which will
+     * each be processed and asynchronically made available in order of processing ticket by later calling
+     * {@link #next()}.
+     *
+     * @param input {@link Iterator} of input to process.
+     * @param shutdownAfterAllSubmitted will call {@link #shutdown(boolean)} after all jobs submitted if {@code true}.
+     * @return {@link Future} representing the work of submitting the inputs to be processed. When the future
+     * is completed all items from the {@code input} {@link Iterator} have been submitted, but some items
+     * may still be queued and processed.
+     */
+    public Future<Void> slurp( Iterator<FROM> input, boolean shutdownAfterAllSubmitted )
+    {
+        return future( () ->
+        {
+            for ( long ticket = 0; input.hasNext(); ticket++ )
+            {
+                submit( ticket, input.next() );
+            }
+            if ( shutdownAfterAllSubmitted )
+            {
+                shutdown( true );
+            }
+            return null;
+        } );
+    }
+
+    /**
+     * Tells this processor that there will be no more submissions and so {@link #next()} will stop blocking,
+     * waiting for new processed results.
+     *
+     * @param awaitAllProcessed if {@code true} will block until all submitted jobs have been processed,
+     * otherwise if {@code false} will return immediately, where processing will still commence and complete.
+     */
+    public void shutdown( boolean awaitAllProcessed )
+    {
+        done = true;
+        executor.shutdown( awaitAllProcessed ? TaskExecutor.SF_AWAIT_ALL_COMPLETED : 0 );
+    }
+
+    /**
+     * @return next processed job (blocking call), or {@code null} if all jobs have been processed
+     * and {@link #shutdown(boolean)} has been called.
+     */
+    public TO next()
+    {
+        while ( !done || processedTicket.get() < submittedTicket.get() || !processed.isEmpty()  )
+        {
+            TO next = processed.poll();
+            if ( next != null )
+            {
+                return next;
+            }
+            park.park( Thread.currentThread() );
+            healthCheck.run();
```

Comment:
Why not use the blocking properties of your `processed` queue here, and in `submit()`?